File: app/src/main/java/com/cray/software/passwords/utils/SharedPrefs.java
package com.cray.software.passwords.utils;

import android.content.Context;
import android.content.SharedPreferences;

import com.google.gson.Gson;

import java.io.File;
import java.io.FileInputStream;
import java.io.FileNotFoundException;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.ObjectInputStream;
import java.io.ObjectOutputStream;
import java.util.Map;

/**
 * Copyright 2016 Nazar Suhovich
 * <p/>
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * <p/>
 * http://www.apache.org/licenses/LICENSE-2.0
 * <p/>
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

abstract class SharedPrefs extends PrefsConstants {

    private static final String TAG = "SharedPrefs";
    private SharedPreferences prefs;
    protected SharedPreferences prefsPassword;

    private SharedPrefs() {
    }

    SharedPrefs(Context context) {
        prefs = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE);
        prefsPassword = context.getSharedPreferences(NEW_APP_PREFS, Context.MODE_PRIVATE);
    }

    void putString(String stringToSave, String value) {
        prefs.edit().putString(stringToSave, value).apply();
    }

    void putInt(String stringToSave, int value) {
        prefs.edit().putInt(stringToSave, value).apply();
    }

    int getInt(String stringToLoad) {
        int x;
        try {
            x = prefs.getInt(stringToLoad, 0);
        } catch (ClassCastException e) {
            try {
                x = Integer.parseInt(prefs.getString(stringToLoad, "4"));
            } catch (ClassCastException e1) {
                x = 0;
            }
        }
        return x;
    }

    void putLong(String stringToSave, long value) {
        prefs.edit().putLong(stringToSave, value).apply();
    }

    long getLong(String stringToLoad) {
        long x;
        try {
            x = prefs.getLong(stringToLoad, 1000);
        } catch (ClassCastException e) {
            x = Long.parseLong(prefs.getString(stringToLoad, "1000"));
        }
        return x;
    }

    void putObject(String key, Object obj) {
        Gson gson = new Gson();
        putString(key, gson.toJson(obj));
    }

    Object getObject(String key, Class<?> classOfT) {
        String json = getString(key);
        Object value = new Gson().fromJson(json, classOfT);
        if (value == null) {
            return new Object();
        }
        return value;
    }

    String getString(String stringToLoad, String def) {
        return prefs.getString(stringToLoad, def);
    }

    String getString(String stringToLoad) {
        return prefs.getString(stringToLoad, null);
    }

    public boolean hasKey(String checkString) {
        return prefs.contains(checkString);
    }

    void putBoolean(String stringToSave, boolean value) {
        prefs.edit().putBoolean(stringToSave, value).apply();
    }

    boolean getBoolean(String stringToLoad) {
        boolean res;
        try {
            res = prefs.getBoolean(stringToLoad, false);
        } catch (ClassCastException e) {
            res = Boolean.parseBoolean(prefs.getString(stringToLoad, "false"));
        }
        return res;
    }

    public boolean saveSharedPreferencesToFile(File dst) {
        boolean res = false;
        ObjectOutputStream output = null;
        try {
            output = new ObjectOutputStream(new FileOutputStream(dst));
            output.writeObject(prefs.getAll());
            res = true;
        } catch (FileNotFoundException e) {
            e.printStackTrace();
        } catch (IOException e) {
            e.printStackTrace();
        } finally {
            try {
                if (output != null) {
                    output.flush();
                    output.close();
                }
            } catch (IOException ex) {
                ex.printStackTrace();
            }
        }
        return res;
    }

    @SuppressWarnings({"unchecked"})
    public boolean loadSharedPreferencesFromFile(File src) {
        boolean res = false;
        ObjectInputStream input = null;
        try {
            input = new ObjectInputStream(new FileInputStream(src));
            SharedPreferences.Editor prefEdit = prefs.edit();
            prefEdit.clear();
            Map<String, ?> entries = (Map<String, ?>) input.readObject();
            for (Map.Entry<String, ?> entry : entries.entrySet()) {
                Object v = entry.getValue();
                String key = entry.getKey();

                if (v instanceof Boolean)
                    prefEdit.putBoolean(key, (Boolean) v);
                else if (v instanceof Float)
                    prefEdit.putFloat(key, (Float) v);
                else if (v instanceof Integer)
                    prefEdit.putInt(key, (Integer) v);
                else if (v instanceof Long)
                    prefEdit.putLong(key, (Long) v);
                else if (v instanceof String)
                    prefEdit.putString(key, ((String) v));
            }
            prefEdit.apply();
            res = true;
        } catch (FileNotFoundException e) {
            e.printStackTrace();
        } catch (IOException e) {
            e.printStackTrace();
        } catch (ClassNotFoundException e) {
            e.printStackTrace();
        } finally {
            try {
                if (input != null) {
                    input.close();
                }
            } catch (IOException ex) {
                ex.printStackTrace();
            }
        }
        return res;
    }
}

File: app/src/main/java/com/cray/software/passwords/utils/SharedPrefs.kt
package com.cray.software.passwords.utils

import android.content.Context
import android.content.SharedPreferences
import java.io.*

/**
 * Copyright 2016 Nazar Suhovich
 *
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

abstract class SharedPrefs(context: Context) : PrefsConstants() {

    private val prefs: SharedPreferences = context.getSharedPreferences(PrefsConstants.PREFS_NAME, Context.MODE_PRIVATE)
    protected var prefsPassword: SharedPreferences = context.getSharedPreferences(PrefsConstants.LOGIN_PREFS, Context.MODE_PRIVATE)


    fun putString(stringToSave: String, value: String) {
        prefs.edit().putString(stringToSave, value).apply()
    }

    fun putInt(stringToSave: String, value: Int) {
        prefs.edit().putInt(stringToSave, value).apply()
    }

    fun getInt(stringToLoad: String): Int {
        var x: Int
        try {
            x = prefs.getInt(stringToLoad, 0)
        } catch (e: ClassCastException) {
            x = 0
        }

        return x
    }

    fun putLong(stringToSave: String, value: Long) {
        prefs.edit().putLong(stringToSave, value).apply()
    }

    fun getLong(stringToLoad: String): Long {
        var x: Long
        try {
            x = prefs.getLong(stringToLoad, 1000)
        } catch (e: ClassCastException) {
            x = 1000
        }

        return x
    }

    fun getString(stringToLoad: String, def: String): String {
        return prefs.getString(stringToLoad, def) ?: def
    }

    fun getString(stringToLoad: String): String {
        return prefs.getString(stringToLoad, "") ?: ""
    }

    fun hasKey(checkString: String): Boolean {
        return prefs.contains(checkString)
    }

    fun putBoolean(stringToSave: String, value: Boolean) {
        prefs.edit().putBoolean(stringToSave, value).apply()
    }

    fun getBoolean(stringToLoad: String): Boolean {
        var res: Boolean
        try {
            res = prefs.getBoolean(stringToLoad, false)
        } catch (e: ClassCastException) {
            res = false
        }

        return res
    }

    fun saveSharedPreferencesToFile(dst: File): Boolean {
        var res = false
        var output: ObjectOutputStream? = null
        try {
            output = ObjectOutputStream(FileOutputStream(dst))
            output.writeObject(prefs.all)
            res = true
        } catch (e: FileNotFoundException) {
            e.printStackTrace()
        } catch (e: IOException) {
            e.printStackTrace()
        } finally {
            try {
                if (output != null) {
                    output.flush()
                    output.close()
                }
            } catch (ex: IOException) {
                ex.printStackTrace()
            }

        }
        return res
    }

    fun loadSharedPreferencesFromFile(src: File): Boolean {
        var res = false
        var input: ObjectInputStream? = null
        try {
            input = ObjectInputStream(FileInputStream(src))
            val prefEdit = prefs.edit()
            prefEdit.clear()
            val entries = input.readObject() as Map<String, *>
            for ((key, v) in entries) {

                if (v is Boolean)
                    prefEdit.putBoolean(key, v)
                else if (v is Float)
                    prefEdit.putFloat(key, v)
                else if (v is Int)
                    prefEdit.putInt(key, v)
                else if (v is Long)
                    prefEdit.putLong(key, v)
                else if (v is String)
                    prefEdit.putString(key, v)
            }
            prefEdit.apply()
            res = true
        } catch (e: FileNotFoundException) {
            e.printStackTrace()
        } catch (e: IOException) {
            e.printStackTrace()
        } catch (e: ClassNotFoundException) {
            e.printStackTrace()
        } finally {
            try {
                input?.close()
            } catch (ex: IOException) {
                ex.printStackTrace()
            }

        }
        return res
    }

    companion object {

        private val TAG = "SharedPrefs"
    }
}